Sources/BSON/BSONDocument.swift

```diff
@@ -174,12 +186,107 @@ public struct BSONDocument {
      */
     public subscript(dynamicMember member: String) -> BSON? {
         get { self[member] }
-        set { fatalError("Unimplemented") }
+        set { self[member] = newValue }
+    }
+
+    /**
+     * Sets a BSON element with the corresponding key
+     * if element.value is nil the element is deleted from the BSON
+     */
+    internal mutating func set(key: String, of value: BSON?) throws {
+        if !self.keySet.contains(key) {
+            guard let value = value else {
+                // no-op: key does not exist and the value is nil
+                return
+            }
+            // appending new key
+            self.keySet.insert(key)
+            self._buffer.moveWriterIndex(to: self.byteLength - 1) // setup to overwrite null terminator
+            let size = BSONDocument.append(key: key, of: value, to: &self._buffer)
+            self._buffer.writeInteger(0, endianness: .little, as: UInt8.self) // add back in our null terminator
+            self.byteLength += size
+            return
+        }
+
+        var iter = BSONDocumentIterator(over: self._buffer)
+
+        guard let (start, end) = iter.findByteRange(for: key) else {
+            throw BSONError.InternalError(message: "Cannot find \(key) to delete")
+        }
+
+        var newBuffer = BSON_ALLOCATOR.buffer(capacity: 0)
+
+        guard
+            let prefix = self._buffer.getBytes(at: 0, length: start),
+            let suffix = self._buffer.getBytes(at: end, length: self.byteLength - end)
+        else {
+            throw BSONError.InternalError(
+                message: "Cannot slice buffer from " +
+                    "0 to len \(start) and from \(end) to len \(self.byteLength - end) : \(self._buffer)"
+            )
+        }
+
+        newBuffer.writeBytes(prefix)
+
+        var newSize = self.byteLength - (end - start)
+        if let value = value {
+            // Overwriting
+            let size = BSONDocument.append(key: key, of: value, to: &newBuffer)
+            newSize += size
+
+            guard newSize <= BSON_MAX_SIZE else {
+                throw BSONError.DocumentTooLargeError(value: value.bsonValue, forKey: key)
+            }
+        } else {
+            // Deleting
+            self.keySet.remove(key)
+        }
+
+        newBuffer.writeBytes(suffix)
+
+        self._buffer = newBuffer
+        self.byteLength = newSize
+        guard self.byteLength == self._buffer.readableBytes else {
+            fatalError("BSONDocument's encoded byte length is \(self.byteLength) however the" +
+                "buffer has \(self._buffer.readableBytes) readable bytes")
+        }
+    }
+
+    /// Appends element to underlying BSON bytes, returns the size of the element appended: type + key + value
+    @discardableResult
+    internal static func append(key: String, of value: BSON, to buffer: inout ByteBuffer) -> Int {
```

Comment:
I think "of" is also kind of funny here, as at the call site it is like "append key a of 1 to buffer". but I will leave you and @patrickfreed to bikeshed that 🙂